### PR TITLE
Add Integer arithmetic

### PR DESCRIPTION
**Pull request by smondet**

### CODE REVIEW

**Comment by smondet** [comment]
Cf. also http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_06_04
and
http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap01.html#tag_01_07_02_01

File: src/lib/EDSL.mli [hunk 1]
```diff
@@ -92,6 +92,35 @@ val string : string -> string t
 val int : int -> int t
 val bool : bool -> bool t
 
+module Integer : sig
+  val to_string : int t -> string t
+  val of_string : string t -> int t
+  val bin_op : int t -> [ `Div | `Minus | `Mult | `Plus | `Mod ] -> int t -> int t
+  val add : int t -> int t -> int t
```
Comment (rleonid) [inline src/lib/EDSL.mli]: I'm curious why you want to expose this function? Is there a standard way in shell to represent this logic specifically, or said differently, is this a shell idiom?

File: src/lib/EDSL.mli [hunk 2]
```diff
@@ -92,6 +92,35 @@ val string : string -> string t
...
+  val (mod) : int t -> int t -> int t
+  val cmp : [ `Eq | `Ge | `Gt | `Le | `Lt | `Ne ] -> int t -> int t -> bool t
+  val eq : int t -> int t -> bool t
+  val ne : int t -> int t -> bool t
+  val lt : int t -> int t -> bool t
+  val le : int t -> int t -> bool t
+  val ge : int t -> int t -> bool t
+  val gt : int t -> int t -> bool t
+  val ( = ) : int t -> int t -> bool t
+  val ( <> ) : int t -> int t -> bool t
```
Comment (rleonid) [inline src/lib/EDSL.mli]: Not necessarily something to add now, but are there structural equality notions in shells? For example the `n`th position of an array?

File: src/lib/EDSL.mli (hunk 1 above)
Comment (rleonid) [inline src/lib/EDSL.mli]: Sorry for the misplacement, this is in reference to `bin_op`.

My usual preference for regular variants in specifying fully formed api's. Can't really imagine any more standard operators that you can define. But you know, just a style quibble, feel free to disregard.

File: src/lib/language.ml [hunk 3]
```diff
@@ -230,6 +265,39 @@ let rec to_shell: type a. _ -> a t -> string =
       Literal.to_shell lit
     | Output_as_string e ->
       sprintf "\"$( { %s ; } | od -t o1 -An -v | tr -d ' \\n' )\"" (continue e)
+    | Int_to_string i ->
+      continue (Output_as_string (Raw_cmd (sprintf "printf '%%d' %s" (continue i))))
+    | String_to_int s ->
+      let var = "tmpxxxxijljlijdeifh" in
+      let value = sprintf "\"$%s\"" var in
+      sprintf " $( %s=$( %s ) ; if [ %s -eq %s ] ; then printf -- %s ; else %s ; fi ; ) "
+        var
```
Comment (rleonid) [inline src/lib/language.ml]: I do not understand, can you document a little bit what is going on here?
I guess specifically when `-eq` will be false when comparing this variable.

File: src/test/main.ml [hunk 4]
```diff
@@ -353,6 +353,80 @@ let tests =
...
+          trybin 1 Integer.(int 2 * (int 22 - int 20) / int 4)
+          &&& 
+          trybin 0 Integer.(int 2 * (int 22 - int 20) / int 5)
+          &&& 
+          trybin 8 Integer.(int 42 / int 5)
+          &&& 
+          trybin 2 Integer.(int 42 mod int 5)
+          &&& 
+          trybin 0 Integer.(int 3000 mod int 3)
+        )
```
Comment (rleonid) [inline src/test/main.ml]: I wonder if there is a mechanism (besides ppx) to lift actually integer literals,  specifically inside of the `Integer`, to obviate all these `int`'s .

**Comment by rleonid** [comment]
Modulo some small comments it looks good.

File: src/lib/EDSL.mli (hunk 1 above)
Comment (smondet) [inline src/lib/EDSL.mli]: For now I prefer leaving the API a bit more "open" to see where it goes.

File: src/lib/EDSL.mli (hunk 2 above)
Comment (smondet) [inline src/lib/EDSL.mli]: POSIX's `sh` does not have arrays.

But we could implement them on top of tmp-files or variables :)

File: src/test/main.ml (hunk 4 above)
Comment (smondet) [inline src/test/main.ml]: No I think you need ppx to override the meaning of int literals.

File: src/lib/EDSL.mli (hunk 2 above)
Comment (rleonid) [inline src/lib/EDSL.mli]: Yep. My original point was in referencing other comparisons, not `int`'s. But agreed that a structural equality for shells is harder to define.

File: src/lib/language.ml (hunk 3 above)
Comment (smondet) [inline src/lib/language.ml]: So the `if [ $v -eq $v ]` checks that the variable `v` contains an integer (i.e., in shell, a string that parses as an int).

http://stackoverflow.com/questions/309745/check-that-a-variable-is-a-number-in-unix-shell#309804

Adding a comment in the code.

**Pull request merged**